src/Compilers/Core/Portable/DiagnosticAnalyzer/ShadowCopyAnalyzerAssemblyLoader.cs

```diff
@@ -11,60 +11,69 @@
 using Roslyn.Utilities;
 using System.Collections.Immutable;
 using System.Reflection;
-
-#if NET
-using System.Runtime.Loader;
-#endif
+using System.Globalization;
 
 namespace Microsoft.CodeAnalysis
 {
-    internal sealed class ShadowCopyAnalyzerAssemblyLoader : AnalyzerAssemblyLoader
+    internal sealed class ShadowCopyAnalyzerPathResolver : IAnalyzerPathResolver
     {
         /// <summary>
         /// The base directory for shadow copies. Each instance of
-        /// <see cref="ShadowCopyAnalyzerAssemblyLoader"/> gets its own
+        /// <see cref="ShadowCopyAnalyzerPathResolver"/> gets its own
         /// subdirectory under this directory. This is also the starting point
         /// for scavenge operations.
         /// </summary>
         private readonly string _baseDirectory;
 
+        /// <summary>
+        /// The name of the shadow directory. This will be the same name as used for the <see cref="_mutex"/>
+        /// field. That allows other instances of this type to coordinate on active and inactive
+        /// shadow copy directories.
+        /// </summary>
+        private readonly string _shadowDirectoryName;
+
+        private readonly string _shadowDirectory;
+
+        /// <summary>
+        /// As long as this mutex is alive, other instances of this type will not try to clean
+        /// up the shadow directory.
+        /// </summary>
+        private readonly Mutex _mutex;
+
         internal readonly Task DeleteLeftoverDirectoriesTask;
 
+        private int _directoryCount;
+
+        private readonly ConcurrentDictionary<string, int> _directoryMap = new();
+
         /// <summary>
-        /// The directory where this instance of <see cref="ShadowCopyAnalyzerAssemblyLoader"/>
-        /// will shadow-copy assemblies, and the mutex created to mark that the owner of it is still active.
+        /// Map from the shadow assembly path to the original assembly path
```

Comment:
is the shadow assembly path the "real path"?  consider uniform naming across all these components if so.